app/download_test.go

```diff
@@ -0,0 +1,75 @@
+package app
+
+import (
+	"fmt"
+	"github.com/mattermost/mattermost-server/v5/model"
+	"github.com/stretchr/testify/require"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+)
+
+func TestDownloadFromURL(t *testing.T) {
+	th := Setup(t).InitBasic()
+	defer th.TearDown()
+
+	app := th.App
+	app.Config().PluginSettings.AllowInsecureDownloadUrl = model.NewBool(true)
+
+	retries := 0
+
+	mux := http.NewServeMux()
+	mux.HandleFunc("/succeeds-after-retry", func(w http.ResponseWriter, r *http.Request) {
+		if retries < 2 {
+			http.Error(w, "Request Timed out", http.StatusGatewayTimeout)
+			retries++
+			return
+		}
+
+		_, _ = w.Write([]byte("Your request is successful."))
+	})
+
+	mux.HandleFunc("/fails-forever", func(w http.ResponseWriter, r *http.Request) {
+		http.Error(w, "This would fail forever", http.StatusInternalServerError)
+	})
+
+	testServer := httptest.NewServer(mux)
+
+	type args struct {
+		downloadURL string
+	}
+
+	tests := []struct {
+		name    string
+		args    args
+		wantErr bool
+	}{
+		{
+			name: "Should succeed after two retries",
+			args: args{
+				downloadURL: fmt.Sprintf("%s/succeeds-after-retry", testServer.URL),
+			},
+			wantErr: false,
+		},
+		{
+			name: "Should not retry forever",
+			args: args{
+				downloadURL: fmt.Sprintf("%s/fails-forever", testServer.URL),
+			},
+			wantErr: true,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			retries = 0
```

Comment:
This is probably okay, but a bit risky. Ideally, it would be good to have a struct type with localized state containing the retries variable. And then with each iteration, we call a method to reset the state. Having a shared variable where setting it to 0 affects something quite high above in the code is error-prone, but maybe okay since it's a test. I'd suggest adding a comment to that effect.